grove/connectors/fleetdm/api.py

```diff
@@ -0,0 +1,181 @@
+        self.jmespath_queries = jmespath_queries
+        self.api_uri = api_uri
+
+            
+    def _get(
+        self,
+        url: str,
+        params: Dict[str, Any],
+    ) -> HTTPResponse:
+        """A GET wrapper to handle retries for the caller.
+
+        :param url: URL to perform the HTTP GET against.
+        :param params: HTTP parameters to add to the request.
+
+        :raises RateLimitException: A rate limit was encountered.
+        :raises RequestFailedException: An HTTP request failed.
+
+        :return: HTTP Response object containing the headers and body of a response.
+        """
+        while True:
+            try:
+                response = requests.get(url, headers=self.headers, params=params)
+                response.raise_for_status()
+                break
+            except requests.exceptions.RequestException as err:
+                # Retry on rate-limit, but only if requested.
+                if getattr(err.response, "status_code", None) == 429:
+                    self.logger.warning("Rate-limit was exceeded during request")
+                    if self.retry:
+                        time.sleep(int(err.response.headers.get("Retry-After", "1")))
+                        continue
+                    else:
+                        raise RateLimitException(err)
+
+                raise RequestFailedException(err)
+        return HTTPResponse(headers=response.headers, body=response.json())
+
+    def get_hosts(
+        self,
+        params: Dict[str, Any],
+        jmespath_queries: str,
+        api_uri: str,
+        cursor: Optional[str],
+    ) -> AuditLogEntries:
+        """Fetches a list of hosts which match the provided filters.
+        :param params: get parameters json dict from https://fleetdm.com/docs/rest-api/rest-api#list-hosts
+        :param jmespath_queries: jmespath query to filter JSON response as specified at https://jmespath.org/
+        :param cursor: The cursor passed from host_logs.py that includes the last update date of systems in Grove
+        :return: AuditLogEntries object containing a pagination cursor, and log entries.
+        """
+
+        # Ensuring that the parameters include the correct After, Order Key, and Order Direction
+        # for the pagination and cursor to work correctly
+        params["after"] = cursor
+        params["order_key"] = "software_updated_at"
+        params["order_direction"] = "asc"
+
+        # See psf/requests issue #2651 for why we can happily pass in None values and
+        # not have the request key added to the URI.
+        result = self._get(
+            f"{api_uri}/api/v1/fleet/hosts",
+            params,
+        )
+
+        filteredResults = []
+        # FleetDM returns an empty hosts array if there's no more pages of results,
+        # so swap this for a None object
+        # Otherwise, grab the last seen update date/time to use in the next page of dates
+        if int(len((result.body.get("hosts",[])))) == 0:
```

Comment:
We shouldn't need the additional cast here. We could also just use a "falsy" on the empty list, but checking the length is likely better :)

```suggestion
        if len(result.body.get("hosts", [])) == 0:
```